### PR TITLE
Export FScanner parser to ZScript as ScriptScanner

### PR DESCRIPTION
**Pull request by Cacodemon345**
This PR exposes the internal `FScanner` parser to ZScript as `ScriptScanner`.

`ScriptScanner`s are created with `Create` functions, and destroyed by calling `Destroy` (after which it must not be used). `SetPrependMessage` is used to set the message that will be prepended to the displayed error messages in case of failure. The rest of the exported functions work unchanged from their native counterparts.

### CODE REVIEW

File: src/common/engine/sc_man.cpp [hunk 1]
```diff
@@ -1383,4 +1387,307 @@ int ParseHex(const char* hex, FScriptPosition* sc)
...
+}
+
+DEFINE_ACTION_FUNCTION(_ScriptScanner, Destroy)
+{
+	PARAM_SELF_STRUCT_PROLOGUE(FScanner);
+	self->Close();
+	delete self;
+	return 0;
+}
+
```
Comment (RicardoLuis0) [inline src/common/engine/sc_man.cpp:1443]: .

**Review by RicardoLuis0 — CHANGES_REQUESTED** [review]

File: src/common/engine/sc_man.cpp (hunk 1 above)
Comment (RicardoLuis0) [inline src/common/engine/sc_man.cpp:1443]: .

Comment (Cacodemon345) [reply src/common/engine/sc_man.cpp]: Done.

File: src/common/engine/sc_man.cpp [hunk 2]
```diff
@@ -1383,4 +1387,307 @@ int ParseHex(const char* hex, FScriptPosition* sc)
...
+DEFINE_ACTION_FUNCTION(_ScriptScanner, CreateFromString)
+{
+	PARAM_PROLOGUE;
+	PARAM_STRING(name);
+	PARAM_STRING_VAL(script);
+
+	auto sc = new FScanner;
+	sc->OpenString(name.GetChars(), script);
+	ACTION_RETURN_POINTER(sc);
+}
```
[the remark this answers is not in the document]
Comment (Cacodemon345) [reply src/common/engine/sc_man.cpp]: Done.

**Review by coelckers — CHANGES_REQUESTED** [review]
Please rewrite this to not use multiple inheritance. A lot of ZScripts inner workings circumvent standard C++ logic that ensure type safety and this may easily end up trashing the pointer to such an object because it may not point to where you expect it anymore.

**Comment by Cacodemon345** [comment]
Done as requested.

**Comment by nashmuhandes** [comment]
Please also provide a working example mod that showcases the usage of this feature.

**Comment by Cacodemon345** [comment]
[scriptparserexample.zip](https://github.com/ZDoom/gzdoom/files/13639574/scriptparserexample.zip)
[demoscpt.txt](https://github.com/ZDoom/gzdoom/files/13639577/demoscpt.txt)

Here's one. Run GZDoom with the ZIP file and the uploaded `demoscpt.txt` lump and open any kind of map. It will print `1337` to console.

[demoscpt.txt](https://github.com/ZDoom/gzdoom/files/13639602/demoscpt.txt)

Since this example uses MustGetStringName function, it will abort to the console with the second one.

**Comment by coelckers** [comment]
I merged this manually and changed it to use direct native calls for most functions

**Pull request closed without merging**